src/systemctl/systemctl-list-units.c

```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
         return machine ? strjoin(machine, ":", unit) : strdup(unit);
 }
 
-static int output_units_list(const UnitInfo *unit_infos, size_t c) {
+typedef struct {
+        char ** units;
+        size_t num_units;
+}SliceUnits;
+
+/* Function to insert or update units under slice unit in the hash map */
+static int insert_units_under_slice(const char *slice_name, char **units, size_t num_units) {
+        SliceUnits *slice_units = NULL;
+        int r;
+
+        /* Initialize the hashmap */
+        if (!slice_unit_map) {
+                slice_unit_map = hashmap_new(&string_hash_ops);
+                if (!slice_unit_map)
+                        return log_oom();
+        }
+        /* Allocate memory for array of unit names for a slice */
+        slice_units = new(SliceUnits, sizeof(SliceUnits));
+        if (!slice_units) {
+                mfree(slice_units);
+                return log_oom();
+        }
+
+        slice_units->units = new(char *, sizeof(char *) * num_units);
+        if (!slice_units->units) {
+                mfree(slice_units->units);
+                return log_oom();
+        }
+
+        /* Copy units names into the array */
+        for (size_t i = 0; i < num_units; i++) {
+                slice_units->units[i] = strdup(units[i]);
+                if (!slice_units->units[i]) {
+                        for (size_t j = 0; j < i; j++)
+                                mfree(slice_units->units[j]);
+                        mfree(slice_units->units);
+                        mfree(slice_units);
+                        return log_oom();
+                }
+        }
+        slice_units->num_units = num_units;
+
+        /* Insert the slice name and corresponding units into the hashmap */
+        r = hashmap_put(slice_unit_map, slice_name, slice_units);
+        if (r < 0) {
+                for (size_t i = 0; i < num_units; i++)
+                        mfree(slice_units->units[i]);
+                mfree(slice_units->units);
+                mfree(slice_units);
+                return log_oom();
+        }
+        return r;
+}
+
+/* Function to retrieve units under a slice from the hash map */
+static SliceUnits *get_units_under_slice(const char *slice_name) {
+        SliceUnits *slice_units;
+
+        assert(slice_name);
+        if (!slice_unit_map)
+                return NULL;
```

Comment:
hashmap_get() is happy if you pass NULL as first arg. It treats that like an empty hashmap